dsp/modules/colbertv2.py

```diff
@@ -74,3 +75,118 @@ def colbertv2_post_request_v2_wrapped(*args, **kwargs):
+
+        from colbert import Indexer
+        from colbert.infra import Run, RunConfig
+        with Run().context(RunConfig(nranks=self.colbert_config.nranks, experiment=self.colbert_config.experiment)):  
+            indexer = Indexer(checkpoint=self.checkpoint, config=self.colbert_config)
+            indexer.index(name=self.colbert_config.index_name, collection=self.passages, overwrite=True)
+
+    def get_index(self):
+        try:
+            import colbert
+        except ImportError:
+            print("Colbert not found. Please check your installation or install the module using pip install colbert-ai[faiss-gpu,torch].")
+
+        from colbert import Searcher
+        from colbert.infra import Run, RunConfig
+        
+        with Run().context(RunConfig(experiment=self.colbert_config.experiment)):
+            searcher = Searcher(index=self.colbert_config.index_name, collection=self.passages)
+        return searcher
+    
+    def __call__(self,query:str,k:int=7,**kwargs):
+        import torch
+        
+        if kwargs.get("filtered_pids"):
+            filtered_pids = kwargs.get("filtered_pids")
+            assert type(filtered_pids) == List[int], "The filtered pids should be a list of integers"
+            device = "cuda" if torch.cuda.is_available() else "cpu"
+            results = self.searcher.search(
+                query,
+                #Number of passages to receive
+                k=k, 
+                #Passing the filter function of relevant 
+                filter_fn=lambda pids: torch.tensor(
+                    [pid for pid in pids if pid in filtered_pids],dtype=torch.int32).to(device))
+        else:
+            searcher_results = self.searcher.search(query, k=k)
+        results = []
+        for pid,rank,score in zip(*searcher_results):
+            results.append(dotdict({'long_text':self.searcher.collection[pid],'score':score,'pid':pid}))
+        return results
+
+class ColBERTv2RerankerLocal:
+    
+    def __init__(self,colbert_config=None,checkpoint:str='bert-base-uncased'):
+        try:
+            import colbert
+        except ImportError:
+            print("Colbert not found. Please check your installation or install the module using pip install colbert-ai[faiss-gpu,torch].")
+        """_summary_
+
+        Args:
+            checkpoint_name (str, optional): checkpoint for embeddings. Defaults to 'bert-base-uncased'.
+            colbert_config (ColBERTConfig, optional): Colbert config. Defaults to ColBERTConfig().
+        """
+        self.colbert_config = colbert_config
+        self.checkpoint_name = checkpoint
+        self.colbert_config.checkpoint = checkpoint
+
+    # def __call__(self, *args: Any, **kwargs: Any) -> Any:
+    #     return self.forward(*args, **kwargs)
+
+    def __call__(self,query:str,passages:List[str]=[]):
+        import numpy as np
+        from colbert.modeling.colbert import ColBERT
```

Comment:
I think users may use Colbert as a retriever and as a reranker separately, hence it would be helpful that we check the import errors separately. Please let me know if I am right in my assumption.